### PR TITLE
Set up s3 file upload

### PR DESCRIPTION
**Pull request by jason17h**
### Notion ticket link

[Set up S3 upload](https://www.notion.so/uwblueprintexecs/086c1b564dbf478faa649a8880e34b81?v=9d3f8002a0d1477ba395616dfc082af7&p=73e58b47c00c457a96e5592ffd71ac24)

### Implementation description

Implementing function to upload a file to S3.

### Steps to test

Called the function using a sample file as input and confirmed on the AWS console that the file was properly uploaded.

### What should reviewers focus on?

-

### Checklist

-   [x] My PR name is descriptive and in imperative tense
-   [x] I have run the linter
-   [x] I have requested a review from the PL, and/or other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: utils/uploadToS3.ts
Comment (feng-eric) [inline utils/uploadToS3.ts]: Add comment to document function

File: utils/uploadToS3.ts
Comment (feng-eric) [inline utils/uploadToS3.ts]: I think `new aws.S3()` should be exported to this file so we don't create new s3 instances

File: utils/uploadToS3.ts
Comment (mahadzaryab1) [inline utils/uploadToS3.ts]: return error rather than logging

File: utils/uploadToS3.ts
Comment (mahadzaryab1) [inline utils/uploadToS3.ts]: return success object rather than logging

File: utils/uploadToS3.ts
Comment (mahadzaryab1) [inline utils/uploadToS3.ts]: can you add an example of how this function can/should be called

**Review by mahadzaryab1 — CHANGES_REQUESTED** [review]
Added a few comments. Another thing to consider is that your file is called `uploadToS3.ts` but I think it should maybe just be `s3.ts` or `s3Utility.ts` so that any s3 utility functions can be added rather than just the upload.

File: utils/uploadToS3.ts
Comment (jason17h) [reply utils/uploadToS3.ts]: see https://github.com/uwblueprint/social-diversity-for-children/pull/11/commits/5652c943c9e3b0fb96cab393fd2dfe0a93275e68

File: utils/uploadToS3.ts
Comment (jason17h) [reply utils/uploadToS3.ts]: I assume you meant exported _from_ this file? If so, see https://github.com/uwblueprint/social-diversity-for-children/pull/11/commits/d683e84f081d2625d8dce69dbf3c462f5be123c8
Otherwise, let me know what you meant

Comment (feng-eric) [reply utils/uploadToS3.ts]: After #12 is merged in, perhaps you could set up & export  `s3` object in `aws/index.js`.

File: utils/aws/s3.ts
Comment (feng-eric) [inline utils/aws/s3.ts]: I don't think this code needs to be in this function here. It can be set outside it since there isn't a need to update the config each time we upload to s3. I think you can have this in `aws.index.js`. See my other comment!

File: utils/uploadToS3.ts
Comment (feng-eric) [reply utils/uploadToS3.ts]: Actually, I think it might be okay to merge yours in first. Perhaps you can create an `index.js` file where you set up everything from aws. See `aws/index.js` in https://github.com/uwblueprint/social-diversity-for-children/pull/12/files and https://github.com/uwblueprint/dancefest-web/blob/staging/aws/index.js

File: utils/aws/s3.ts
Comment (mahadzaryab1) [inline utils/aws/s3.ts]: @feng-eric - why does `S3` need to be exported? is it going to be used outside this file?

File: utils/aws/s3.ts
Comment (mahadzaryab1) [inline utils/aws/s3.ts]: the function shouldn't be returning void - it should maybe return some type of response object that indicates if the upload was successful, error messages, and potentially the data object

File: utils/aws/s3.ts
Comment (mahadzaryab1) [inline utils/aws/s3.ts]: Use jsdoc syntax for function documentation - see [here](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html)

File: utils/aws/s3.ts
Comment (mahadzaryab1) [inline utils/aws/s3.ts]: Have you tried this to make sure it works? it might be helpful to test this if you haven't already on some test endpoint.

**Review by mahadzaryab1 — CHANGES_REQUESTED** [review]
Added a few more comments!

File: utils/aws/s3.ts
Comment (feng-eric) [reply utils/aws/s3.ts]: My initial comment was saying we can initialize and export s3 object from `aws/index.js`. We can also just initialize in here outside of the function instead too. Either one works.

File: utils/aws/s3.ts
Comment (feng-eric) [reply utils/aws/s3.ts]: I think it would be most useful to return the response from the s3 upload.

File: utils/aws/s3.ts
Comment (feng-eric) [reply utils/aws/s3.ts]: I didn't mean export it out of this file

File: utils/aws/s3.ts
Comment (brandonwong368) [reply utils/aws/s3.ts]: Yeah, we called the function in an endpoint and got a sample text file to upload

File: utils/uploadToS3.ts
Comment (jason17h) [reply utils/uploadToS3.ts]: Since the success object is in the callback function, can it be returned from the calling function?

File: utils/aws/s3.ts
Comment (jason17h) [reply utils/aws/s3.ts]: Is this what you had in mind? https://github.com/uwblueprint/social-diversity-for-children/pull/11/commits/4156ac1af5483819aca96900b2314dc72f6a486e

File: utils/uploadToS3.ts
Comment (jason17h) [reply utils/uploadToS3.ts]: See https://github.com/uwblueprint/social-diversity-for-children/pull/11/commits/4156ac1af5483819aca96900b2314dc72f6a486e

**Review by mahadzaryab1 — CHANGES_REQUESTED** [review]
Added a few more comments

File: yarn.lock
Comment (brandonwong368) [inline yarn.lock]: I was trying to resolve the conflicts with staging and I think I removed a dependency by accident @mahadzaryab1 @feng-eric, do you know which aws-sdk, buffer, and ieee versions we are supposed to keep

File: utils/aws/s3.ts
Comment (jason17h) [reply utils/aws/s3.ts]: Updated it so that it now returns a promise which the client can work with: https://github.com/uwblueprint/social-diversity-for-children/pull/11/commits/8798d6abefc54c21c27187d32cbade845b641d21

File: package.json
Comment (feng-eric) [inline package.json:47]: How come this was removed?

Comment (feng-eric) [reply package.json:47]: Already talked to @brandonwong368 about this but let's also remove `package-lock.json`

File: aws/index.ts
Comment (feng-eric) [inline aws/index.ts:9]: Is this required?

**Review by feng-eric — APPROVED** [review]
LGTM! Before merging in, let's get @mahadzaryab1 to take a look too and also if you guys could address the comment I put.

File: aws/index.ts
Comment (brandonwong368) [reply aws/index.ts:9]: It's not required, I tested the function with an API endpoint without `signatureVersion: "v4",` and the upload worked fine. 

I took this from the AWS documentation: The Signature Version 2 or Signature Version 4 signing process is used to authenticate your Amazon S3 API requests. Signing requests enables Amazon S3 to identify who is sending the request and protects your requests from bad actors. https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingAWSSDK.html

It seems like removing it could have a security issue later on? Not sure, I won't remove it until @mahadzaryab1  takes a look

Comment (mahadzaryab1) [reply aws/index.ts:9]: I think this is good to keep due to the reasons that @brandonwong368 suggested.

File: utils/aws/s3.ts
Comment (mahadzaryab1) [inline utils/aws/s3.ts]: This function actually isn't returning a promise anymore because you're resolving the promise within the function itself. It seems to either return `data` or `err` which or of two different types and you generally want to avoid doing that in Typescript. You have two options here: 
- Instead of resolving the promise within the function itself, you can return it from the function and let the caller resolve it. 
- Make a custom s3 response object similar to [this](https://github.com/uwblueprint/social-diversity-for-children/blob/staging/src/models/response.ts#L6-L9) and then the caller can check to see if the error field is null or not. If it's not null there's an error and if it is then there's no error and the caller can check the data field.  

The above two suggestions can be done in separate PR if you like but for now it's probably good to change the return type from `Promise<unknown>` to just `unknown`. cc @feng-eric for thoughts

File: utils/aws/s3.ts
Comment (mahadzaryab1) [inline utils/aws/s3.ts]: nit: change comment because the bucket doesn't correspond to `AWS_BUCKET_NAME` anymore from the perspective of this file

**Review by mahadzaryab1 — CHANGES_REQUESTED** [review]
Overall, looks great - nice work!! I just highlighted one thing that needs to be fixed and we should be good to go after!

File: utils/aws/s3.ts
Comment (feng-eric) [reply utils/aws/s3.ts]: I think the second approach with using a custom s3 response object seems good. We can do it in a different PR.

File: utils/aws/s3.ts
Comment (mahadzaryab1) [reply utils/aws/s3.ts]: did you do this ?

**Review by mahadzaryab1 — APPROVED** [review]
LGTM - just fix the comment on the function! Great work!!

**Pull request merged**
